src/v/bytes/iobuf.h

```diff
@@ -394,40 +396,31 @@ iobuf::try_copy_append(const char* ptr, size_t size) {
         return;
     }
 
-    if (available_bytes() > 0) {
-        if (_frags.back().is_empty()) {
-            pop_back();
-        } else {
-            _frags.back().trim();
-        }
+    if (unlikely(available_bytes() > 0 && _frags.back().is_empty())) {
+        pop_back();
     }
     append(std::make_unique<fragment>(std::move(b)));
 }
 
 /// appends the contents of buffer; might pack values into existing space
 inline void iobuf::append(iobuf&& o) {
-    for (auto& f : o._frags) {
-        auto fsize = f.size();
-
-        if (unlikely(!fsize)) {
-            continue;
-        }
+    while (!o._frags.empty()) {
+        fragment* f = &o._frags.front();
+        o._frags.pop_front();
 
-        if (try_copy_append(f.get(), fsize)) {
+        auto fsize = f->size();
+        if (!fsize || try_copy_append(f->get(), fsize)) {
+            details::dispose_io_fragment(f);
             continue;
         }
 
-        if (available_bytes() > 0) {
-            if (_frags.back().is_empty()) {
-                pop_back();
-            } else {
-                _frags.back().trim();
-            }
+        if (unlikely(!_frags.empty() && _frags.back().is_empty())) {
+            pop_back();
         }
 
-        append(std::move(f).unoptimized_release());
+        append(f);
```

Comment:
Note that since we're not trimming `f` here `test_next_chunk_allocation_append_iobuf` will fail as is since the appended data will now map to 322 fragments instead of 323. If we trim `f` the test will pass and the behavior will be identical to what it is today. However, I really don't think that behavior is optimal since if we are taking ownership of the fragments we may as well use any free space rather than just trimming it away. Thoughts?

Reply:
Brandon do you know why we were trimming before here? Seems weird.

Reply:
I don't believe the trimming was ever intentional. However, before this PR appending an existing `io_fragment` to an `iobuf` would entail; 
- The underlying `temporary_buffer` in the `io_fragment` to be appended was shared up to the used bytes. This is basically what is acting as a trim.
- A new `io_fragment` was allocated and given the shared `temporary_buffer`.
- This new fragment was appended to the `iobuf`